salt/cloud/__init__.py

```diff
@@ -2083,128 +2083,142 @@ def run_map(self, dmap):
 
         # Now sort the create list based on dependencies
         create_list = sorted(six.iteritems(dmap['create']), key=lambda x: x[1]['level'])
+        full_map = dmap['create'].copy()
+        full_map.update(dmap['existing'])
+        possible_master_list = sorted(six.iteritems(full_map), key=lambda x: x[1]['level'])
         output = {}
         if self.opts['parallel']:
             parallel_data = []
         master_name = None
         master_minion_name = None
         master_host = None
         master_finger = None
-        try:
-            master_name, master_profile = next((
-                (name, profile) for name, profile in create_list
-                if profile.get('make_master', False) is True
-            ))
-            master_minion_name = master_name
-            log.debug('Creating new master \'{0}\''.format(master_name))
-            if salt.config.get_cloud_config_value(
-                'deploy',
-                master_profile,
-                self.opts
-            ) is False:
-                raise SaltCloudSystemExit(
-                    'Cannot proceed with \'make_master\' when salt deployment '
-                    'is disabled(ex: --no-deploy).'
-                )
-
-            # Generate the master keys
-            log.debug(
-                'Generating master keys for \'{0[name]}\''.format(master_profile)
-            )
-            priv, pub = salt.utils.cloud.gen_keys(
-                salt.config.get_cloud_config_value(
-                    'keysize',
+        for name, profile in possible_master_list:
+            if profile.get('make_master', False) is True:
+                master_name = name
+                master_profile = profile
+
+        if master_name:
+            # If the master already exists, get the host
+            if master_name not in dmap['create']:
+                client = salt.client.get_local_client()
+                master_host = client.cmd(master_name, 'grains.get', ['host'])[master_name]
```

Comment:
I do not like this, we should do a query of the cloud provider to get this information instead of checking the grains.

There is always the possibility that the machine is not connected to this master.

Reply:
Do you perhaps have some pointers on how to use salt cloud to query internally? I couldn't find a nice feature for that.

Reply:
You are already inside of salt-cloud, just use `self.query()` and find the instance.